### PR TITLE
feat: New module `avm-res-keyvault-vault`

### PR DESCRIPTION
**Pull request by eriqua**
## Description

Adding avm-res-keyvault-vault AVM module

### CODE REVIEW

File: avm/res/key-vault/vault/tests/e2e/defaults/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    lock: {}
+    roleAssignments: []
+    diagnosticSettings: []
+    privateEndpoints: []
+    accessPolicies: []
+    secrets: {}
+    keys: []
+    tags: {}
+  }
+}
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/defaults/main.test.bicep:59]:
```suggestion
@batchSize(1)
module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
  scope: resourceGroup
  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
  params: {
    name: '${namePrefix}${serviceShort}002'
    location: location
    // Only for testing purposes
    enablePurgeProtection: false
    // Workaround for PSRule
    lock: {}
    roleAssignments: []
    diagnosticSettings: []
    privateEndpoints: []
    accessPolicies: []
    secrets: {}
    keys: []
    tags: {}
  }
}]
```

File: avm/res/key-vault/vault/main.bicep [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+      name: sku
+      family: 'A'
+    }
+    networkAcls: !empty(networkAcls) ? {
+      bypass: networkAcls.?bypass
+      defaultAction: networkAcls.?defaultAction
+      virtualNetworkRules: networkAcls.?virtualNetworkRules ?? []
+      ipRules: networkAcls.?ipRules ?? []
+    } : null
+    publicNetworkAccess: !empty(publicNetworkAccess) ? any(publicNetworkAccess) : (!empty(privateEndpoints) && empty(networkAcls) ? 'Disabled' : null)
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/main.bicep:162]: This actually fails for I think CongnitiveServices. It wants 'Disabled' if I'm not mistaken, but the same rule funnily enough does not exist for kvlt.

File: avm/res/key-vault/vault/tests/e2e/max/main.test.bicep [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+  }
+}
+
+// ============== //
+// Test Execution //
+// ============== //
+
+module testDeployment '../../../main.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/max/main.test.bicep:64]:
```suggestion
@batchSize(1)
module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
  scope: resourceGroup
  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
```

File: avm/res/key-vault/vault/tests/e2e/max/main.test.bicep [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+      ]
+    }
+    softDeleteRetentionInDays: 7
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
+  }
+}
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/max/main.test.bicep:237]:
```suggestion
}]
```

File: avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  name: '${uniqueString(deployment().name, location)}-nestedDependencies'
+  params: {
+    virtualNetworkName: 'dep-${namePrefix}-vnet-${serviceShort}'
+    location: location
+  }
+}
+
+// Diagnostics
+// ===========
+module diagnosticDependencies '../../../../../../utilities/e2e-template-assets/templates/diagnostic.dependencies.bicep' = {
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep:45]: Did we come to a conclusion if we want to keep these resources in tests that are just about a specific feature? Happy to, just wondering if we came to a decision or just leave it to the module owner.

File: avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  }
+}
+
+// ============== //
+// Test Execution //
+// ============== //
+
+module testDeployment '../../../main.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep:63]:
```suggestion
@batchSize(1)
module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
  scope: resourceGroup
  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
```

File: avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    ]
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
+    // Only for testing purposes
+    enablePurgeProtection: false
+  }
+}
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep:88]:
```suggestion
}]
```

File: avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  }
+}
+
+// ============== //
+// Test Execution //
+// ============== //
+
+module testDeployment '../../../main.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}'
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep:64]:
```suggestion
@batchSize(1)
module testDeployment '../../../main.bicep' = [for iteration in [ 'init', 'idem' ]: {
  scope: resourceGroup
  name: '${uniqueString(deployment().name, location)}-test-${serviceShort}-${iteration}'
```

File: avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    tags: {
+      'hidden-title': 'This is visible in the resource name'
+      Environment: 'Non-Prod'
+      Role: 'DeploymentValidation'
+    }
+    // Workaround for PSRule
+    roleAssignments: []
+    accessPolicies: []
+  }
+}
```
Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep:149]:
```suggestion
}]
```

Comment (AlexanderSehr) [inline avm/res/key-vault/vault/tests/e2e/waf-aligned/main.test.bicep:149]:
```suggestion
}]
```

File: avm/res/key-vault/vault/main.bicep (hunk 2 above)
Comment (AlexanderSehr) [reply avm/res/key-vault/vault/main.bicep:162]: We'll revisit this again later (cc: @AlexanderSehr, @eriqua)

File: avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep (hunk 5 above)
Comment (eriqua) [reply avm/res/key-vault/vault/tests/e2e/private-endpoint/main.test.bicep:45]: agreed and removed

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**